### PR TITLE
[Bitbucket] Use workspaces instead of deprecated/removed teams path

### PR DESCRIPTION
**Pull request by MartinHerren**
Crude fix to allow to correctly list workspaces for bitbucket cloud (https://bitbucket.org) and so run a pipeline.
Last year they removed a bunch of deprecated APIs and replaced them with new ones.

User management is still broken (maybe even more with this PR), it doesn't retrieve correct user info like email for example.

Unit tests are not yet aligned with my changes. Will try to fix that.

Got a server running locally connected to bitbucket.org, can privately provide the url if someone needs testing.

Discovered a 'feature' of bitbucket: oauth2 tokens are created at 'workspace' level but are not scoped to a specific workspace, so anyone with a bitbucket.org account can login into woodpecker. This makes it important not to allow user registration, if not possible use filters to allow only the desired workspaces in order to prevent anyone to abuse your server and runners to build their own pipelines.

I also renamed a few functions containing 'Team' to 'Workspace'. I don't know the convention here, if inside a forge you prefer to keep the 'woodpecker terminology' or the 'server specific terminology'. There is no notion of 'Team' in bitbucket cloud. Only workspaces and projects. From what i understood a repo is part of a workspace and can be part of a project. A project only serves to apply some global git settings (like merge strategies) to a group of repos. User management is done at workspace and repo level. Also the git url only contains the workspace and repo.

I'm not a Go developer yet so still learning...

Signed-off-by: Martin Herren <martin.herren@gmail.com>

### CODE REVIEW

**Comment by 6543** [comment]
what whould help, is if you can post raw json webhook data ... so we can adjust webhook parsing :)

**Comment by MartinHerren** [comment]
> what whould help, is if you can post raw json webhook data ... so we can adjust webhook parsing :)

@6543 I'm not sure i understand exactly what you need

**Comment by 6543** [comment]
Please dont force-push, pulls get squash-merged anyway.

Just merge master into it and push

**Comment by 6543** [comment]
With webhook request body, i mean the raw data bitbucket send via http(s) to woodpecker-server on events happening on bitbucket

It's data encoded as json in plain text in an http(s)-request in the so called body

**Comment by 6543** [comment]
Gitea/github/gitlab lets you obtain it in the reposection if webhooks are enabled and already got send by cliking on an webhook entry to see details

**Comment by MartinHerren** [comment]
Rebased on master and fixed the build of the tests. Now there are still 2 tests not passing.

**Comment by 6543** [comment]
@MartinHerren please dont rebase :/

-> https://github.com/woodpecker-ci/woodpecker/pull/1166#issuecomment-1239713102

**Comment by 6543** [comment]
pull title and description will be used on merge for the final commit

**Comment by MartinHerren** [comment]
Sorry i'm used to rebase my wip branches to keep all my branches and my history clean... automatic reflex. Will try not to rebase.

**Comment by MartinHerren** [comment]
> 



> Gitea/github/gitlab lets you obtain it in the repo-settings if webhooks are enabled and already got send by cliking on an webhook entry to see details

Ok, think i found a way to activate history logging of webhooks. Just need to create a dummy project on my test account to generate the webhook data.

**Comment by MartinHerren** [comment]
As .json files seem not to be allowed as attachement by github... here the webhook data  on a push event:

```
{
  "push": {
    "changes": [
      {
        "old": {
          "name": "master",
          "target": {
            "type": "commit",
            "hash": "a51241ae1f00cbe728930db48e890b18fd527f99",
            "date": "2022-08-17T15:24:29+00:00",
            "author": {
              "type": "author",
              "raw": "Martin Herren <martin.herren@ecorobotix.com>",
              "user": {
                "display_name": "Martin Herren",
                "links": {
                  "self": {
                    "href": "https://api.bitbucket.org/2.0/users/%7B69cc59f2-706b-4a9c-b99c-eac2ace320da%7D"
                  },
                  "avatar": {
                    "href": "https://secure.gravatar.com/avatar/7b2e50690b4ab7bb9e1db18ea3b8ae95?d=https%3A%2F%2Favatar-management--avatars.us-west-2.prod.public.atl-paas.net%2Finitials%2FMH-5.png"
                  },
                  "html": {
                    "href": "https://bitbucket.org/%7B69cc59f2-706b-4a9c-b99c-eac2ace320da%7D/"
                  }
                },
                "type": "user",
                "uuid": "{69cc59f2-706b-4a9c-b99c-eac2ace320da}",
                "account_id": "5d286e857133f10c17e026cb",
                "nickname": "Martin Herren"
              }
            },
            "message": "Add test .woodpecker.yml\n",
            "summary": {
              "type": "rendered",
              "raw": "Add test .woodpecker.yml\n",
              "markup": "markdown",
              "html": "<p>Add test .woodpecker.yml</p>"
            },
            "links": {
              "self": {
                "href": "https://api.bitbucket.org/2.0/repositories/martinherren1984/publictestrepo/commit/a51241ae1f00cbe728930db48e890b18fd527f99"
              },
              "html": {
                "href": "https://bitbucket.org/martinherren1984/publictestrepo/commits/a51241ae1f00cbe728930db48e890b18fd527f99"
              }
            },
            "parents": [],
            "rendered": {},
            "properties": {}
          },
          "links": {
            "self": {
              "href": "https://api.bitbucket.org/2.0/repositories/martinherren1984/publictestrepo/refs/branches/master"
            },
            "commits": {
              "href": "https://api.bitbucket.org/2.0/repositories/martinherren1984/publictestrepo/commits/master"
            },
            "html": {
              "href": "https://bitbucket.org/martinherren1984/publictestrepo/branch/master"
            }
          },
          "type": "branch",
          "merge_strategies": [
            "merge_commit",
            "squash",
            "fast_forward"
          ],
          "default_merge_strategy": "merge_commit"
        },
        "new": {
          "name": "master",
          "target": {
            "type": "commit",
            "hash": "c14c1bb05dfb1fdcdf06b31485fff61b0ea44277",
            "date": "2022-09-07T20:19:25+00:00",
            "author": {
              "type": "author",
              "raw": "Martin Herren <martin.herren@gmail.com>",
              "user": {
                "display_name": "Martin Herren",
                "links": {
                  "self": {
                    "href": "https://api.bitbucket.org/2.0/users/%7Bc5a0d676-fd27-4bd4-ac69-a7540d7b495b%7D"
                  },
                  "avatar": {
                    "href": "https://secure.gravatar.com/avatar/37de364488b2ec474b5458ca86442bbb?d=https%3A%2F%2Favatar-management--avatars.us-west-2.prod.public.atl-paas.net%2Finitials%2FMH-2.png"
                  },
                  "html": {
                    "href": "https://bitbucket.org/%7Bc5a0d676-fd27-4bd4-ac69-a7540d7b495b%7D/"
                  }
                },
                "type": "user",
                "uuid": "{c5a0d676-fd27-4bd4-ac69-a7540d7b495b}",
                "account_id": "5cf8e3a9678ca90f8e7cc8a8",
                "nickname": "Martin Herren"
              }
            },
            "message": "a\n",
            "summary": {
              "type": "rendered",
              "raw": "a\n",
              "markup": "markdown",
              "html": "<p>a</p>"
            },
            "links": {
              "self": {
                "href": "https://api.bitbucket.org/2.0/repositories/martinherren1984/publictestrepo/commit/c14c1bb05dfb1fdcdf06b31485fff61b0ea44277"
              },
              "html": {
                "href": "https://bitbucket.org/martinherren1984/publictestrepo/commits/c14c1bb05dfb1fdcdf06b31485fff61b0ea44277"
              }
            },
            "parents": [
              {
                "type": "commit",
                "hash": "a51241ae1f00cbe728930db48e890b18fd527f99",
                "links": {
                  "self": {
                    "href": "https://api.bitbucket.org/2.0/repositories/martinherren1984/publictestrepo/commit/a51241ae1f00cbe728930db48e890b18fd527f99"
                  },
                  "html": {
                    "href": "https://bitbucket.org/martinherren1984/publictestrepo/commits/a51241ae1f00cbe728930db48e890b18fd527f99"
                  }
                }
              }
            ],
            "rendered": {},
            "properties": {}
          },
          "links": {
            "self": {
              "href": "https://api.bitbucket.org/2.0/repositories/martinherren1984/publictestrepo/refs/branches/master"
            },
            "commits": {
              "href": "https://api.bitbucket.org/2.0/repositories/martinherren1984/publictestrepo/commits/master"
            },
            "html": {
              "href": "https://bitbucket.org/martinherren1984/publictestrepo/branch/master"
            }
          },
          "type": "branch",
          "merge_strategies": [
            "merge_commit",
            "squash",
            "fast_forward"
          ],
          "default_merge_strategy": "merge_commit"
        },
        "truncated": false,
        "created": false,
        "forced": false,
        "closed": false,
        "links": {
          "commits": {
            "href": "https://api.bitbucket.org/2.0/repositories/martinherren1984/publictestrepo/commits?include=c14c1bb05dfb1fdcdf06b31485fff61b0ea44277&exclude=a51241ae1f00cbe728930db48e890b18fd527f99"
          },
          "diff": {
            "href": "https://api.bitbucket.org/2.0/repositories/martinherren1984/publictestrepo/diff/c14c1bb05dfb1fdcdf06b31485fff61b0ea44277..a51241ae1f00cbe728930db48e890b18fd527f99"
          },
          "html": {
            "href": "https://bitbucket.org/martinherren1984/publictestrepo/branches/compare/c14c1bb05dfb1fdcdf06b31485fff61b0ea44277..a51241ae1f00cbe728930db48e890b18fd527f99"
          }
        },
        "commits": [
          {
            "type": "commit",
            "hash": "c14c1bb05dfb1fdcdf06b31485fff61b0ea44277",
            "date": "2022-09-07T20:19:25+00:00",
            "author": {
              "type": "author",
              "raw": "Martin Herren <martin.herren@gmail.com>",
              "user": {
                "display_name": "Martin Herren",
                "links": {
                  "self": {
                    "href": "https://api.bitbucket.org/2.0/users/%7Bc5a0d676-fd27-4bd4-ac69-a7540d7b495b%7D"
                  },
                  "avatar": {
                    "href": "https://secure.gravatar.com/avatar/37de364488b2ec474b5458ca86442bbb?d=https%3A%2F%2Favatar-management--avatars.us-west-2.prod.public.atl-paas.net%2Finitials%2FMH-2.png"
                  },
                  "html": {
                    "href": "https://bitbucket.org/%7Bc5a0d676-fd27-4bd4-ac69-a7540d7b495b%7D/"
                  }
                },
                "type": "user",
                "uuid": "{c5a0d676-fd27-4bd4-ac69-a7540d7b495b}",
                "account_id": "5cf8e3a9678ca90f8e7cc8a8",
                "nickname": "Martin Herren"
              }
            },
            "message": "a\n",
            "summary": {
              "type": "rendered",
              "raw": "a\n",
              "markup": "markdown",
              "html": "<p>a</p>"
            },
            "links": {
              "self": {
                "href": "https://api.bitbucket.org/2.0/repositories/martinherren1984/publictestrepo/commit/c14c1bb05dfb1fdcdf06b31485fff61b0ea44277"
              },
              "html": {
                "href": "https://bitbucket.org/martinherren1984/publictestrepo/commits/c14c1bb05dfb1fdcdf06b31485fff61b0ea44277"
              },
              "diff": {
                "href": "https://api.bitbucket.org/2.0/repositories/martinherren1984/publictestrepo/diff/c14c1bb05dfb1fdcdf06b31485fff61b0ea44277"
              },
              "approve": {
                "href": "https://api.bitbucket.org/2.0/repositories/martinherren1984/publictestrepo/commit/c14c1bb05dfb1fdcdf06b31485fff61b0ea44277/approve"
              },
              "comments": {
                "href": "https://api.bitbucket.org/2.0/repositories/martinherren1984/publictestrepo/commit/c14c1bb05dfb1fdcdf06b31485fff61b0ea44277/comments"
              },
              "statuses": {
                "href": "https://api.bitbucket.org/2.0/repositories/martinherren1984/publictestrepo/commit/c14c1bb05dfb1fdcdf06b31485fff61b0ea44277/statuses"
              },
              "patch": {
                "href": "https://api.bitbucket.org/2.0/repositories/martinherren1984/publictestrepo/patch/c14c1bb05dfb1fdcdf06b31485fff61b0ea44277"
              }
            },
            "parents": [
              {
                "type": "commit",
                "hash": "a51241ae1f00cbe728930db48e890b18fd527f99",
                "links": {
                  "self": {
                    "href": "https://api.bitbucket.org/2.0/repositories/martinherren1984/publictestrepo/commit/a51241ae1f00cbe728930db48e890b18fd527f99"
                  },
                  "html": {
                    "href": "https://bitbucket.org/martinherren1984/publictestrepo/commits/a51241ae1f00cbe728930db48e890b18fd527f99"
                  }
                }
              }
            ],
            "rendered": {},
            "properties": {}
          }
        ]
      }
    ]
  },
  "repository": {
    "type": "repository",
    "full_name": "martinherren1984/publictestrepo",
    "links": {
      "self": {
        "href": "https://api.bitbucket.org/2.0/repositories/martinherren1984/publictestrepo"
      },
      "html": {
        "href": "https://bitbucket.org/martinherren1984/publictestrepo"
      },
      "avatar": {
        "href": "https://bytebucket.org/ravatar/%7B898477b2-a080-4089-b385-597a783db392%7D?ts=default"
      }
    },
    "name": "PublicTestRepo",
    "scm": "git",
    "website": null,
    "owner": {
      "display_name": "Martin Herren",
      "links": {
        "self": {
          "href": "https://api.bitbucket.org/2.0/users/%7Bc5a0d676-fd27-4bd4-ac69-a7540d7b495b%7D"
        },
        "avatar": {
          "href": "https://secure.gravatar.com/avatar/37de364488b2ec474b5458ca86442bbb?d=https%3A%2F%2Favatar-management--avatars.us-west-2.prod.public.atl-paas.net%2Finitials%2FMH-2.png"
        },
        "html": {
          "href": "https://bitbucket.org/%7Bc5a0d676-fd27-4bd4-ac69-a7540d7b495b%7D/"
        }
      },
      "type": "user",
      "uuid": "{c5a0d676-fd27-4bd4-ac69-a7540d7b495b}",
      "account_id": "5cf8e3a9678ca90f8e7cc8a8",
      "nickname": "Martin Herren"
    },
    "workspace": {
      "type": "workspace",
      "uuid": "{c5a0d676-fd27-4bd4-ac69-a7540d7b495b}",
      "name": "Martin Herren",
      "slug": "martinherren1984",
      "links": {
        "avatar": {
          "href": "https://bitbucket.org/workspaces/martinherren1984/avatar/?ts=1658761964"
        },
        "html": {
          "href": "https://bitbucket.org/martinherren1984/"
        },
        "self": {
          "href": "https://api.bitbucket.org/2.0/workspaces/martinherren1984"
        }
      }
    },
    "is_private": false,
    "project": {
      "type": "project",
      "key": "PUB",
      "uuid": "{2cede481-f59e-49ec-88d0-a85629b7925d}",
      "name": "PublicTestProject",
      "links": {
        "self": {
          "href": "https://api.bitbucket.org/2.0/workspaces/martinherren1984/projects/PUB"
        },
        "html": {
          "href": "https://bitbucket.org/martinherren1984/workspace/projects/PUB"
        },
        "avatar": {
          "href": "https://bitbucket.org/account/user/martinherren1984/projects/PUB/avatar/32?ts=1658768453"
        }
      }
    },
    "uuid": "{898477b2-a080-4089-b385-597a783db392}"
  },
  "actor": {
    "display_name": "Martin Herren",
    "links": {
      "self": {
        "href": "https://api.bitbucket.org/2.0/users/%7Bc5a0d676-fd27-4bd4-ac69-a7540d7b495b%7D"
      },
      "avatar": {
        "href": "https://secure.gravatar.com/avatar/37de364488b2ec474b5458ca86442bbb?d=https%3A%2F%2Favatar-management--avatars.us-west-2.prod.public.atl-paas.net%2Finitials%2FMH-2.png"
      },
      "html": {
        "href": "https://bitbucket.org/%7Bc5a0d676-fd27-4bd4-ac69-a7540d7b495b%7D/"
      }
    },
    "type": "user",
    "uuid": "{c5a0d676-fd27-4bd4-ac69-a7540d7b495b}",
    "account_id": "5cf8e3a9678ca90f8e7cc8a8",
    "nickname": "Martin Herren"
  }
}
```

It is from a dummy repo in a dummy account i created for testing...

File: server/remote/bitbucket/fixtures/handler.go [hunk 1]
```diff
@@ -42,6 +43,11 @@ func Handler() http.Handler {
 	return e
 }
 
+func getWorkspace(c *gin.Context) {
+	// role, _ := c.Params.Get("role")
```
Comment (6543) [inline server/remote/bitbucket/fixtures/handler.go:47]: this is needed to let tests pass again (mock get workspaces call)

File: server/remote/bitbucket/fixtures/hooks.go [hunk 2]
```diff
@@ -17,48 +18,333 @@ package fixtures
 const HookPush = `
 {
   "actor": {
-    "username": "emmap1",
```
Comment (6543) [inline server/remote/bitbucket/fixtures/hooks.go:20]: your demo data will now be used to test agains :)

File: server/remote/bitbucket/fixtures/handler.go [hunk 3]
```diff
@@ -27,6 +27,7 @@ func Handler() http.Handler {
 	gin.SetMode(gin.TestMode)
 
 	e := gin.New()
+	e.GET("/2.0/workspaces/", getWorkspace)
```
Comment (MartinHerren) [inline server/remote/bitbucket/fixtures/handler.go:30]: As of https://developer.atlassian.com/cloud/bitbucket/bitbucket-api-teams-deprecation/ this new path should completely replace the "/2.0/teams/" path below which doesn't exist anymore.

Comment (6543) [reply server/remote/bitbucket/fixtures/handler.go:30]: Then jut delete the old one ...

File: server/remote/bitbucket/fixtures/handler.go (hunk 1 above)
Comment (6543) [reply server/remote/bitbucket/fixtures/handler.go:47]: (I mean the todo below has to be solved)

**Comment by aabril2me** [comment]
+1 to this PR.
 
My golang skills are limited at the moment, I'd love to contribute.

**Comment by MartinHerren** [comment]
> +1 to this PR.
> 
> My golang skills are limited at the moment, I'd love to contribute.

Yes, time to finish this PR ! Will attack the testing part again.

File: server/remote/bitbucket/fixtures/handler.go (hunk 1 above)
Comment (MartinHerren) [reply server/remote/bitbucket/fixtures/handler.go]: Solved to the best of my little go knowledge. Mostly adapted the old getUserTeams() method and returning a correct workspaces object. Mostly unsure about the switch clause and whether the role should be used. At least tests are passing but no idea if the tests are correct now.

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
So a step closer to workin BB :)

**Comment by MartinHerren** [comment]
> So a step closer to working BB :)

Yeah, thanks ! The next bigger step i see is probably the user management.
Then
- register/handle the other webhooks
- handle multi pipelines
But know it is already possible to build pipelines !

Thanks !

**Comment by aabril2me** [comment]
in my company, we are migrating to woodpecker thanks to this PR (we use Bitbucket). 
Appreciate that @MartinHerren !!

**Comment by 6543** [comment]
ok let's chat at  #1047 ...

**Pull request merged**